src/storage/mvcc/txn.rs

```diff
@@ -423,6 +426,263 @@ impl<S: Snapshot> MvccTxn<S> {
+                    cid,
+                    err: txn::Error::Mvcc(Error::WriteConflict {
+                        start_ts: ts,
+                        conflict_start_ts: write.start_ts,
+                        conflict_commit_ts: commit_ts,
+                        key: key.into_raw()?,
+                        primary: primary.to_vec(),
+                    }),
+                    tag,
+                }));
+            } else if let Result::Err(e) =
+                self.check_data_constraint(is_insert, &write, commit_ts, &key)
+            {
+                return Ok(Some(Msg::FinishedWithErr {
+                    cid,
+                    err: txn::Error::Mvcc(e),
+                    tag,
+                }));
+            }
+        } else if iter.prev(statistics) {
+            let write_key = iter.key(statistics);
+            if Key::is_user_key_eq(write_key, key.as_encoded()) {
+                let write = Write::parse(iter.value(statistics))?;
+                let commit_ts = Key::decode_ts_from(write_key)?;
+                if let Err(e) = self.check_data_constraint(is_insert, &write, commit_ts, &key) {
+                    return Ok(Some(Msg::FinishedWithErr {
+                        cid,
+                        err: txn::Error::Mvcc(e),
+                        tag,
+                    }));
+                }
+            }
+        }
+        Ok(None)
+    }
+
+    pub fn batch_prewrite<Sched: MsgScheduler>(
+        &mut self,
+        cid: u64,
+        commands: &[Command],
+        ids: &mut Vec<u64>,
+        scheduler: &Sched,
+    ) -> Result<u64> {
+        let mut map = BTreeMap::new();
+        let len = commands.len();
+        let tag = commands[0].tag();
+        let mut rows = 0;
+        // initialize mutation to txn mapping
+        for i in 0..len {
+            if let Command::Prewrite { mutations, .. } = &commands[i] {
+                let mutation_len = mutations.len();
+                for j in 0..mutation_len {
+                    map.insert(mutations[j].key().clone().into_encoded(), (i, j));
+                }
+            }
+        }
+        let (lower, lower_idx) = map.iter().next().unwrap();
+        let (upper, _) = map.iter().next_back().unwrap();
+        let mut min_key = Key::from_encoded(lower.to_vec());
+        let mut upper = upper.to_vec();
+        *upper.last_mut().unwrap() += 1;
+        let max_key = Key::from_encoded(upper);
+        let mut min_idx = lower_idx;
+        let mut min_ts = if let Command::Prewrite { start_ts, .. } = commands[min_idx.0] {
+            start_ts
+        } else {
+            0
+        };
+        let mut write_iter = self
+            .reader
+            .new_write_cursor(&min_key, min_ts, &max_key, u64::max_value())
+            .unwrap();
+        let mut lock_iter = self.reader.new_lock_cursor(&min_key, &max_key).unwrap();
+        write_iter.seek(
+            &min_key.clone().append_ts(min_ts),
```

Comment:
should be max_ts

Reply:
I checked that `Seek` returns the min key >= seeked key. Here it should give us the first write exceeding start-ts if any. Btw in latest patch I use cursor.prev() to check data constraints on earlier writes.

Reply:
> I checked that `Seek` returns the min key >= seeked key. Here it should give us the first write exceeding start-ts if any. Btw in latest patch I use cursor.prev() to check data constraints on earlier writes.

ts is encoded in reverse order.